mobile_verifier/src/boosting_oracles/mod.rs

```diff
@@ -1,162 +1,94 @@
-            Some((_, &[1])) => Ok(Assignment::A),
-            Some((_, &[0])) => Ok(Assignment::B),
-            None => Ok(Assignment::C),
-            Some((_, other)) => {
-                anyhow::bail!("unexpected urbanization disktree data: {cell:?} {other:?}")
-            }
-        }
+impl HexAssignment for Assignment {
+    fn assignment(&self, _cell: hextree::Cell) -> anyhow::Result<Assignment> {
+        Ok(*self)
     }
+}
 
-    fn footfall_assignment(&self, cell: hextree::Cell) -> anyhow::Result<Assignment> {
-        let Some((_, vals)) = self.footfall.get(cell)? else {
-            return Ok(Assignment::C);
-        };
-
-        match vals {
-            &[x] if x >= 1 => Ok(Assignment::A),
-            &[0] => Ok(Assignment::B),
-            other => anyhow::bail!("unexpected footfall disktree data: {cell:?} {other:?}"),
-        }
+#[derive(derive_builder::Builder)]
+#[builder(pattern = "owned")]
+pub struct HexBoostData<Foot, Land, Urban> {
+    pub footfall: Foot,
+    pub landtype: Land,
+    pub urbanization: Urban,
+}
+impl<F, L, U> HexBoostData<F, L, U> {
+    pub fn builder() -> HexBoostDataBuilder<F, L, U> {
+        HexBoostDataBuilder::default()
     }
+}
 
-    fn landtype_assignment(&self, cell: hextree::Cell) -> anyhow::Result<Assignment> {
-        let Some((_, vals)) = self.landtype.get(cell)? else {
-            return Ok(Assignment::C);
-        };
-
-        anyhow::ensure!(
-            vals.len() == 1,
-            "unexpected landtype disktree data: {cell:?} {vals:?}"
-        );
-
-        let cover = Landtype::try_from(vals[0])?;
-        Ok(cover.into())
+impl<Foot, Land, Urban> HexBoostData<Foot, Land, Urban>
+where
+    Foot: DataSet,
+    Land: DataSet,
+    Urban: DataSet,
+{
+    pub fn is_ready(&self) -> bool {
+        self.urbanization.is_ready() && self.footfall.is_ready() && self.landtype.is_ready()
     }
 }
 
-impl From<Landtype> for Assignment {
-    fn from(value: Landtype) -> Self {
-        match value {
-            Landtype::Built => Assignment::A,
-            //
-            Landtype::Tree => Assignment::B,
-            Landtype::Shrub => Assignment::B,
-            Landtype::Grass => Assignment::B,
-            //
-            Landtype::Bare => Assignment::C,
-            Landtype::Crop => Assignment::C,
-            Landtype::Frozen => Assignment::C,
-            Landtype::Water => Assignment::C,
-            Landtype::Wet => Assignment::C,
-            Landtype::Mangrove => Assignment::C,
-            Landtype::Moss => Assignment::C,
-        }
+impl<Foot, Land, Urban> HexBoostData<Foot, Land, Urban>
+where
+    Foot: HexAssignment,
+    Land: HexAssignment,
+    Urban: HexAssignment,
+{
+    pub async fn assignments(&self, cell: hextree::Cell) -> anyhow::Result<HexAssignments> {
+        HexAssignments::builder(cell)
+            .footfall(&self.footfall)
+            .landtype(&self.landtype)
+            .urbanized(&self.urbanization)
+            .build()
```

Comment:
Since the locks are gone from the underlying data, this function no longer needs to be `async`, and can be constructed in place removing the need for the a Builder.

```suggestion
    pub fn assignments(&self, cell: hextree::Cell) -> anyhow::Result<HexAssignments> {
        let footfall = self.footfall.assignment(cell)?;
        let landtype = self.landtype.assignment(cell)?;
        let urbanized = self.urbanization.assignment(cell)?;
        Ok(HexAssignments {
            footfall,
            landtype,
            urbanized,
        })
```